File: samples/vulnerable_app.js
/**
 * Sample vulnerable Node.js application for testing the security agent
 * Contains intentional security vulnerabilities
 */

const express = require('express');
const mysql = require('mysql');
const crypto = require('crypto');

const app = express();

// Security Issue 1: Hardcoded credentials
const API_KEY = "sk-1234567890abcdef";
const DB_PASSWORD = "root123";

// Security Issue 2: Insecure database connection
const db = mysql.createConnection({
    host: 'localhost',
    user: 'root',
    password: DB_PASSWORD,
    database: 'myapp'
});

class AuthService {
    constructor() {
        this.sessions = new Map();
    }

    // Security Issue 3: SQL Injection vulnerability
    authenticateUser(username, password, callback) {
        const query = `SELECT * FROM users WHERE username = '${username}' AND password = '${password}'`;
        db.query(query, (error, results) => {
            if (error) {
                callback(error, null);
                return;
            }
            callback(null, results.length > 0);
        });
    }

    // Security Issue 4: More SQL injection
    getUserById(userId) {
        return new Promise((resolve, reject) => {
            const query = "SELECT * FROM users WHERE id = " + userId;
            db.query(query, (error, results) => {
                if (error) reject(error);
                else resolve(results[0]);
            });
        });
    }

    // Security Issue 5: Weak session token generation
    generateSessionToken() {
        return Math.random().toString(36).substring(2);
    }

    // Security Issue 6: Insecure password hashing
    hashPassword(password) {
        return crypto.createHash('md5').update(password).digest('hex');
    }
}

// Security Issue 7: XSS vulnerability in template
app.get('/user/:id', async (req, res) => {
    const userId = req.params.id;
    const authService = new AuthService();
    
    try {
        const user = await authService.getUserById(userId);
        if (user) {
            // Direct HTML injection without escaping
            const html = `<h1>Welcome ${user.name}!</h1><p>Bio: ${user.bio}</p>`;
            res.send(html);
        } else {
            res.status(404).send('User not found');
        }
    } catch (error) {
        res.status(500).send('Database error');
    }
});

// Security Issue 8: Command injection
app.get('/logs/:filename', (req, res) => {
    const filename = req.params.filename;
    const { exec } = require('child_process');
    
    // Dangerous command execution
    exec(`cat /var/log/${filename}`, (error, stdout, stderr) => {
        if (error) {
            res.status(500).send('Error reading log file');
            return;
        }
        res.send(`<pre>${stdout}</pre>`);
    });
});

// Security Issue 9: Eval usage
app.post('/calculate', (req, res) => {
    const expression = req.body.expression;
    try {
        // Never use eval with user input!
        const result = eval(expression);
        res.json({ result: result });
    } catch (error) {
        res.status(400).json({ error: 'Invalid expression' });
    }
});

// Security Issue 10: Path traversal
app.get('/files/:filename', (req, res) => {
    const filename = req.params.filename;
    const fs = require('fs');
    const path = `/app/uploads/${filename}`;
    
    try {
        const content = fs.readFileSync(path, 'utf8');
        res.send(content);
    } catch (error) {
        res.status(404).send('File not found');
    }
});

// Security Issue 11: Unsafe innerHTML usage (client-side)
function updateUserProfile(userData) {
    const profileDiv = document.getElementById('profile');
    // XSS vulnerability
    profileDiv.innerHTML = `
        <h2>${userData.name}</h2>
        <p>Email: ${userData.email}</p>
        <p>About: ${userData.about}</p>
    `;
}

// Security Issue 12: No input validation
app.post('/login', (req, res) => {
    const { username, password } = req.body;
    const authService = new AuthService();
    
    authService.authenticateUser(username, password, (error, isValid) => {
        if (error) {
            res.status(500).json({ error: 'Database error' });
            return;
        }
        
        if (isValid) {
            const token = authService.generateSessionToken();
            res.json({ success: true, token: token });
        } else {
            res.status(401).json({ error: 'Invalid credentials' });
        }
    });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

File: samples/vulnerable_app.py
#!/usr/bin/env python3
"""
Sample vulnerable web application for testing the security agent
Contains intentional security vulnerabilities
"""

import os
import sqlite3
import hashlib
import random
from flask import Flask, request, render_template_string

app = Flask(__name__)

# Security Issue 1: Hardcoded secret
SECRET_KEY = "super_secret_api_key_12345"
DATABASE_PASSWORD = "admin123"

class UserManager:
    def __init__(self):
        self.db_path = "users.db"
        self.init_database()
    
    def init_database(self):
        conn = sqlite3.connect(self.db_path)
        # Security Issue 2: SQL Injection vulnerability
        conn.execute("""
            CREATE TABLE IF NOT EXISTS users (
                id INTEGER PRIMARY KEY,
                username TEXT,
                password TEXT,
                email TEXT
            )
        """)
        conn.close()
    
    def authenticate_user(self, username, password):
        conn = sqlite3.connect(self.db_path)
        # Security Issue 3: SQL Injection - direct string formatting
        query = f"SELECT * FROM users WHERE username='{username}' AND password='{password}'"
        cursor = conn.execute(query)
        user = cursor.fetchone()
        conn.close()
        return user is not None
    
    def get_user_by_id(self, user_id):
        conn = sqlite3.connect(self.db_path)
        # Security Issue 4: More SQL injection
        query = "SELECT * FROM users WHERE id=" + str(user_id)
        cursor = conn.execute(query)
        user = cursor.fetchone()
        conn.close()
        return user
    
    def hash_password(self, password):
        # Security Issue 5: Weak hashing (MD5)
        return hashlib.md5(password.encode()).hexdigest()
    
    def generate_session_token(self):
        # Security Issue 6: Insecure random
        return str(random.random() * 1000000)

@app.route('/login', methods=['POST'])
def login():
    username = request.form.get('username')
    password = request.form.get('password')
    
    user_manager = UserManager()
    if user_manager.authenticate_user(username, password):
        token = user_manager.generate_session_token()
        return f"Login successful! Token: {token}"
    else:
        return "Login failed"

@app.route('/profile')
def profile():
    user_id = request.args.get('id')
    user_manager = UserManager()
    user = user_manager.get_user_by_id(user_id)
    
    if user:
        # Security Issue 7: XSS vulnerability
        template = f"<h1>Welcome {user[1]}!</h1><p>Email: {user[3]}</p>"
        return render_template_string(template)
    else:
        return "User not found"

@app.route('/search')
def search():
    query = request.args.get('q', '')
    # Security Issue 8: Command injection
    import subprocess
    result = subprocess.run(f"grep -r '{query}' /tmp/", shell=True, capture_output=True)
    return f"Search results: {result.stdout.decode()}"

def read_config_file(filename):
    # Security Issue 9: Path traversal
    config_path = f"/app/config/{filename}"
    try:
        with open(config_path, 'r') as f:
            return f.read()
    except:
        return None

if __name__ == '__main__':
    # Security Issue 10: Debug mode in production
    app.run(debug=True, host='0.0.0.0')
